components/builder-depot/src/server.rs

```diff
@@ -461,32 +473,34 @@ fn download_latest_origin_encryption_key(req: &mut Request) -> IronResult<Respon
         OriginPublicEncryptionKey,
     >(req, &request) {
         Ok(key) => key,
-        Err(err) => return Ok(render_net_error(&err)),
+        Err(err) => {
+            if err.get_code() == ErrCode::ENTITY_NOT_FOUND {
+                generate_origin_encryption_keys(&origin, req)?;
+                return download_latest_origin_encryption_key(req);
+            } else {
+                return Ok(render_net_error(&err));
+            }
+        }
     };
 
     let xfilename = format!("{}-{}.pub", key.get_name(), key.get_revision());
     download_content_as_file(key.get_body(), xfilename)
 }
 
-fn generate_origin_encryption_keys(req: &mut Request) -> IronResult<Response> {
-    debug!("Generate Origin Encryption Keys {:?}", req);
-    match get_param(req, "origin") {
-        Some(origin) => {
-            if !check_origin_access(req, &origin).unwrap_or(false) {
-                return Ok(Response::with(status::Forbidden));
-            }
+fn generate_origin_encryption_keys(origin: &str, req: &mut Request) -> IronResult<Response> {
```

Comment:
I'm not sure we even need a generation function at this level at all with the new strategy - this could all get collapsed inside of the get_origin_encryption_keys in originsrv (maybe renamed to get_or_create_origin_encrytpion_keys w/ a GetOrCreateOriginEncryptionKeys message).